scipy/interpolate/_fitpack_repro.py

```diff
@@ -0,0 +1,823 @@
+""" Replicate FITPACK's logic, in bits and pieces.
+"""
+
+
+import warnings
+import operator
+import math
+import numpy as np
+
+from scipy.interpolate._bsplines import (
+    _not_a_knot, make_lsq_spline, make_interp_spline, BSpline,
+    fpcheck, PackedMatrix, _qr_reduce, fpback, _lsq_solve_qr
+)
+
+#    cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+#    c  part 1: determination of the number of knots and their position     c
+#    c  **************************************************************      c
+#
+# https://github.com/scipy/scipy/blob/maintenance/1.11.x/scipy/interpolate/fitpack/fpcurf.f#L31
+
+# Hardcoded in curfit.f
+TOL = 0.001
+MAXIT = 20
+
+
+def _get_residuals(x, y, t, k, w):
+ #   from scipy.interpolate._bsplines import make_lsq_spline
+    # FITPACK has (w*(spl(x)-y))**2; make_lsq_spline has w*(spl(x)-y)**2
+
+    w2 = w**2
+    # XXX: reuse the QR solver from gh-19753
+    spl = make_lsq_spline(x, y, w=w2, t=t, k=k)
+    residuals = w2 * (spl(x) - y)**2
+    return residuals
+
+
+def _add_knot(x, y, t, k, w):
+    """Add a new knot.
+
+    (Approximately) replicate FITPACK's logic:
+      1. split the `x` array into knot intervals, ``t(j+k) <= x(i) <= t(j+k+1)``
+      2. find the interval with the maximum sum of residuals
+      3. insert a new knot into the middle of that interval.
+
+    NB: a new knot is in fact an `x` value at the middle of the interval.
+    So *the knots are a subset of `x`*.
+
+    This routine is an analog of
+    https://github.com/scipy/scipy/blob/v1.11.4/scipy/interpolate/fitpack/fpcurf.f#L190-L215
+    (cf _split function)
+
+    and https://github.com/scipy/scipy/blob/v1.11.4/scipy/interpolate/fitpack/fpknot.f
+    """
+    fparts, ix = _split(x, y, t, k, w)
+
+    ### redo
+    assert all(ix == np.searchsorted(x, t[k:-k]))
+
+    # find the interval with max fparts and non-zero number of x values inside
+    idx_max = -101
+    fpart_max = -1e100
+    for i in range(len(fparts)):
+        if ix[i+1] - ix[i] > 1 and fparts[i] > fpart_max:
+            idx_max = i
+            fpart_max = fparts[i]
+
+    if idx_max == -101:
+        raise ValueError("Internal error, please report it to SciPy developers.")
```

Comment:
Looks like it may be "ok" that this line is hard to hit (not covered by tests at the moment).

Reply:
Yup, this should never be hit (and if it is, we'd need to fix the cause and flip it back to "never be hit").